### PR TITLE
 FTS should record transition of primary to SYNC

### PR DESCRIPTION
**Pull request by ashwinstar**
Short snippets of commit messages for ease

commit b6f802aa60bdc80735d7e870f9b47e1cc5813f9e

    Do not fast exit SyncRepWaitForLSN() based on SyncStandbysDefined()

    By setting up synchronized replication between primary and standby, commits on
    primary should be blocked by function `SyncRepWaitForLSN()` until its effects
    replicated to the standby. `SyncRepWaitForLSN()` was fast exiting if the
    `SyncStandbysDefined()` is false, which means GUC `synchronous_standby_names` is
    empty.

    There was a race condition where the GUC is set but not reflected by some
    backends. However, the `pg_stat_replication` `sync_state` is `sync`, and the
    `state` is `streaming`, which means the commit should block and get replicated
    to the mirror before it returns.

    The change is to NOT do the fast exit based on the GUC, but only rely on
    `WalSndCtl->sync_standbys_defined`.

    Issue was spotted by Asim Praveen during whiteboard discussion.
    Discussion: https://www.postgresql.org/message-id/flat/CABrsG8j3kPD+kbbsx_isEpFvAgaOBNGyGpsqSjQ6L8vwVUaZAQ@mail.gmail.com

commit baf2d0b49f634fb81db8953261c7ba269d6e2358

    Unit tests for FTS state transitions.

commit 875c8d7a3b53423ad9a2b38529d3af244930765a

    FTS correctly handle double fault scenario.

    When primary and mirror are not in sync, cluster is running with no HA for that
    segment. If FTS fails to probe primary as well, currently nothing much can be
    done except retrying. So, in such case FTS should not mark both the segment as
    down as it would be cure is worse than the disease.

commit 57a5387af7e8a2d5e585a6620a124ff1fc17bc22

    FTS should record transition of primary to SYNC.

    If walsender state transitions to streaming, FTS must record transition of same
    on master in catalog. Recording the state to SYNC will enable failing over to
    mirror incase of need.

commit 92a1665922bdc90649490fc1bac0a20ce2eac05a

    Use macros from gp_segment_config.h and delete redundant definition

### CODE REVIEW

File: src/backend/fts/test/fts_test.c [hunk 1]
```diff
@@ -506,20 +538,234 @@ test_probeWalRepPublishUpdate_multiple_segments(void **state)
...
+
+	/* set the mirror down in config */
+	CdbComponentDatabaseInfo *cdbinfo =
+			GetSegmentFromCdbComponentDatabases(
+				cdb_component_dbs, 0, GP_SEGMENT_CONFIGURATION_ROLE_MIRROR);
+	cdbinfo->status = GP_SEGMENT_CONFIGURATION_STATUS_DOWN;
+
+	FtsWalRepInitProbeContext(cdb_component_dbs, &context);
+
+	/* No response received for probe */
```
Comment (asimrp) [inline src/backend/fts/test/fts_test.c]: A more explicit comment will help, such as "No response received from segment 1 (content 0 primary)".

File: src/backend/fts/test/fts_test.c [hunk 2]
```diff
@@ -506,20 +538,234 @@ test_probeWalRepPublishUpdate_multiple_segments(void **state)
...
+		 */
 		unit_test(test_probeWalRepPublishUpdate_for_zero_segment),
 		unit_test(test_probeWalRepPublishUpdate_for_shutdown_requested),
 		unit_test(test_probeWalRepPublishUpdate_for_FtsIsActive_false),
-		unit_test(test_probeWalRepPublishUpdate_no_update),
-		unit_test(test_probeWalRepPublishUpdate_primary_up_to_down),
-		unit_test(test_probeWalRepPublishUpdate_mirror_up_to_down),
-		unit_test(test_probeWalRepPublishUpdate_mirror_down_to_up),
-		unit_test(test_probeWalRepPublishUpdate_multiple_segments)
+		unit_test(test_probeWalRepPublishUpdate_multiple_segments),
```
Comment (asimrp) [inline src/backend/fts/test/fts_test.c]: Nice to see all combinations covered :+1:

File: src/backend/fts/test/fts_test.c [hunk 3]
```diff
@@ -506,20 +538,234 @@ test_probeWalRepPublishUpdate_multiple_segments(void **state)
...
+	/* Probe responded with Mirror Up and SYNC */
+	context.responses[0].result.isPrimaryAlive = true;
+	context.responses[0].result.isMirrorAlive = true;
+	context.responses[0].result.isInSync = true;
+
+	/* we are updating one segment pair */
+	PrimaryOrMirrorWillBeUpdated(1);
+
+	/* primary and mirror both will be updated */
+	MockPrimaryAndMirrorProbeResponse(cdb_component_dbs, &context.responses[0], true, true);
```
Comment (xinzweb) [inline src/backend/fts/test/fts_test.c]: Optional: change the API to actually specify the expected primary/mirror status/mode, other than just do `true/false` for whether the catalog updated or not.

File: src/backend/fts/test/fts_test.c [hunk 4]
```diff
@@ -506,20 +538,234 @@ test_probeWalRepPublishUpdate_multiple_segments(void **state)
...
+		unit_test(test_PrimayUpMirrorUpSync_to_PrimaryDown),
+
+		unit_test(test_PrimayUpMirrorUpNotInSync_to_PrimayUpMirrorUpSync),
+		unit_test(test_PrimayUpMirrorUpNotInSync_to_PrimayUpMirrorUpNotInSync),
+		unit_test(test_PrimayUpMirrorUpNotInSync_to_PrimaryUpMirrorDownNotInSync),
+		unit_test(test_PrimayUpMirrorUpNotInSync_to_PrimaryDown),
+
+		unit_test(test_PrimaryUpMirrorDownNotInSync_to_PrimayUpMirrorUpSync),
+		unit_test(test_PrimaryUpMirrorDownNotInSync_to_PrimayUpMirrorUpNotInSync),
+		unit_test(test_PrimaryUpMirrorDownNotInSync_to_PrimaryDown)
```
Comment (xinzweb) [inline src/backend/fts/test/fts_test.c]: verify another no-op case: `PrimaryUpMirrorDownNotInSync_to_PrimaryUpMirrorDownNotInSync()`

File: src/backend/fts/fts.c [hunk 5]
```diff
@@ -413,10 +415,13 @@ probeWalRepUpdateConfig(int16 dbid, int16 segindex, bool IsSegmentAlive)
 		histvals[Anum_gp_configuration_history_dbid-1] =
 				Int16GetDatum(dbid);
 		snprintf(desc, sizeof(desc),
-				"FTS: update status for dbid %d with contentid %d to %c",
-				dbid, segindex,
-				IsSegmentAlive ? GP_SEGMENT_CONFIGURATION_STATUS_UP :
-					GP_SEGMENT_CONFIGURATION_STATUS_DOWN);
+				 "FTS: update status for dbid %d with contentid %d to %c and %c",
```
Comment (xinzweb) [inline src/backend/fts/fts.c]: `update status and mode for`

File: src/backend/replication/gp_replication.c [hunk 6]
```diff
@@ -22,9 +22,10 @@
  * True: if any mirror is up.
  * False: if none of the mirrors is up.
  */
-bool IsMirrorUp(void)
+void GetMirrorStatus(bool *IsMirrorUp, bool *IsInSync)
```
[the remark this answers is not in the document]
Comment (xinzweb) [reply src/backend/replication/gp_replication.c]: If there are more than 5 fields, we probably need to create some struct like `MirrorStatus` to encapsulate it. At now, we should be ok with just 2 fields. 

Structure of `ProbeReponse` semantically can contain other information like IO, network, etc. for the primary, which is more than just mirror status.

[the remark this answers is not in the document]
Comment (asimrp) [reply src/backend/replication/gp_replication.c]: Sounds good.

**Pull request merged**